File: clients/go/events.go
package dotprompt

import (
	"context"
	"encoding/json"
	"io"
)

type EventStream struct {
	decoder *json.Decoder
}

func NewEventStream(resp io.Reader) *EventStream {
	return &EventStream{
		decoder: json.NewDecoder(resp),
	}
}

func (es *EventStream) Next() (*DotPromptEvent, error) {
	var event DotPromptEvent
	if err := es.decoder.Decode(&event); err != nil {
		return nil, err
	}
	return &event, nil
}

func CreateEventStream(ctx context.Context, baseURL string, apiKey string) (<-chan DotPromptEvent, error) {
	transport := NewTransport(TransportOptions{
		BaseURL: baseURL,
		APIKey:  apiKey,
	})

	resp, err := transport.stream(ctx, "/api/events")
	if err != nil {
		return nil, err
	}

	ch := make(chan DotPromptEvent, 10)

	go func() {
		defer resp.Body.Close()
		defer close(ch)

		decoder := json.NewDecoder(resp.Body)
		for {
			select {
			case <-ctx.Done():
				return
			default:
			}

			var event DotPromptEvent
			if err := decoder.Decode(&event); err != nil {
				if err == io.EOF {
					return
				}
				return
			}
			ch <- event
		}
	}()

	return ch, nil
}

File: clients/typescript/src/events.ts
import { createParser, type ParsedEvent, type ReconnectInterval } from "eventsource-parser";
import { type DotPromptEvent } from "./models.js";

/**
 * Creates an async generator for receiving SSE events from the dot-prompt API.
 * Uses fetch streaming and eventsource-parser for efficient parsing.
 * 
 * @param url - The /api/events URL.
 * @param apiKey - Optional API key for authorization.
 * @returns AsyncGenerator<DotPromptEvent>
 */
export async function* createEventStream(
  url: string,
  apiKey?: string
): AsyncGenerator<DotPromptEvent> {
  const headers: Record<string, string> = {
    "Accept": "text/event-stream",
  };

  if (apiKey) {
    headers["Authorization"] = `Bearer ${apiKey}`;
  }

  const response = await fetch(url, { headers });

  if (!response.ok) {
    const errorText = await response.text();
    throw new Error(`Failed to establish event stream: ${response.status} ${errorText}`);
  }

  if (response.body === null) {
      throw new Error("Response body is null");
  }

  const reader = response.body.getReader();
  const decoder = new TextDecoder();

  let resolve: (event: DotPromptEvent | null) => void;
  let eventQueue: DotPromptEvent[] = [];
  let isDone = false;

  const parser = createParser((event: ParsedEvent | ReconnectInterval) => {
    if (event.type === "event") {
      try {
        const data = JSON.parse(event.data);
        eventQueue.push(data as DotPromptEvent);
      } catch {
        // Log or handle malformed events?
      }
    }
  });

  try {
    while (true) {
      const { done, value } = await reader.read();
      
      if (done) {
        isDone = true;
        break;
      }

      parser.feed(decoder.decode(value, { stream: true }));

      while (eventQueue.length > 0) {
        yield eventQueue.shift()!;
      }
    }
  } finally {
    reader.releaseLock();
  }
}
